krb5_locator: Allow hostname in kdcinfo files

thalman opened the pull request:
Currently we support only IP addresses in kdcinfo files. We need
to resolv eventual dns name and then we have to iterate trough
list of addresses because hostname can be resolved that way, including
IPv4 and IPv6 addresses.

Resolves:
https://pagure.io/SSSD/sssd/issue/3973

jhrozek commented:
btw I filed https://pagure.io/SSSD/sssd/issue/3973 to track this part of the effort

sumit-bose commented:
@thalman, can you rebase your patch on top of current master?

sumit-bose commented on src/krb5_plugin/sssd_krb5_locator_plugin.c:
```diff
@@ -569,22 +570,32 @@ krb5_error_code sssd_krb5_locator_lookup(void *private_data,
-        PLUGIN_DEBUG("addr[%s:%s] family[%d] socktype[%d]\n",
-                     addr[c].addr, port_str, ai->ai_family, ai->ai_socktype);
+        for (ai_item = ai; ai_item != NULL; ai_item = ai_item->ai_next) {
+            ret = getnameinfo(ai_item->ai_addr, ai_item->ai_addrlen,
```
Since the getnameinfo() result is only needed for the debug output I would suggest to only call it if ctx->debug is true.

sumit-bose commented on src/krb5_plugin/sssd_krb5_locator_plugin.c:
```diff
@@ -569,22 +570,32 @@ krb5_error_code sssd_krb5_locator_lookup(void *private_data,
             return KRB5_PLUGIN_NO_HANDLE;
```
Since there can be multiple addresses and names in the kdcinfo file and actually DNS lookups might be done now I think it might be better to have the possibility not fail completely but just try the next entry.

Just error out to avoid further delays with upcoming DNS lookups (which have a chance to fail as well) is imo a sensible default.

But I wonder what you think about adding another environment variable like e.g. SSSD_KRB5_LOCATOR_IGNORE_DNS_FAILURES to tell the plugin to ignore the error and continue with the next entry?

sumit-bose commented:
Hi Tomas,

please see my inline comments. Additionally please do a rebase not a merge so that your patch is the topmost commit and there is no merge commit. Please add to like to the pagure ticket instead of the design page to the commit message and add a comment to the man page for the plugin which explains the new feature.

bye,
Sumit

thalman commented:
Hi @sumit-bose 

Thank you for the review, I updated PR according comments, Also I find out that one man-page
should be also updated so it is included in PR now too.

Tom

jhrozek commented:
A total nitpick, but you'll probably want to add resolves https://pagure.io/SSSD/sssd/issue/3973 and if you want to point to the design page, I've rendered it at https://docs.pagure.org/SSSD.sssd/design_pages/kdcinfo_multiple_servers.html

sumit-bose commented on src/krb5_plugin/sssd_krb5_locator_plugin.c:
```diff
@@ -581,25 +593,42 @@ krb5_error_code sssd_krb5_locator_lookup(void *private_data,
+                }
+            }
+
+            PLUGIN_DEBUG("addr[%s (%s)] port[%s] family[%d] socktype[%d]\n",
```
Can you move the PLUGIN_DEBUG() inside the if (ctx->debug)-block. It is not needed but would be more readable imo.

sumit-bose commented:
Hi @thalman,

thank you for the new version. It looks good now except the minor nitpick I added inline. Would you mind to fix this as well?

bye,
Sumit

jhrozek commented on src/man/sssd_krb5_locator_plugin.8.xml:
```diff
@@ -94,6 +96,12 @@
+            If the environment variable SSSD_KRB5_LOCATOR_IGNORE_DNS_FAILURES
+            is set to any value plugin will try to resolve all DNS names
+            in kdcinfo file. By default plugin returns KRB5_PLUGIN_NO_HANDLE
+            to the caller immediatelly on first DNS resolving failure.
```
Nitpick: s/immediatelly/immediately/

jhrozek commented on src/man/sssd_krb5_locator_plugin.8.xml:
```diff
@@ -49,12 +49,14 @@
-            number can be added to the end separated with a colon, the IPv6
-            address has to be enclosed in squared brackets in this case as
-            usual. Valid entries are:
+            should contain one or more dns names or IP addresses either in
```
s/dns/DNS/

thalman commented:
@jhrozek, @sumit-bose PR updated

sumit-bose commented:
Thanks, ACK.

jhrozek commented:
master: 63ccbfe
sssd-1-16: 96e4d71
